File: stripe/src/main/java/com/stripe/android/CustomerEphemeralKey.java
package com.stripe.android;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import org.json.JSONException;
import org.json.JSONObject;

final class CustomerEphemeralKey extends EphemeralKey {
    public static final Parcelable.Creator<CustomerEphemeralKey> CREATOR
            = new Parcelable.Creator<CustomerEphemeralKey>() {

        @Override
        public CustomerEphemeralKey createFromParcel(Parcel in) {
            return new CustomerEphemeralKey(in);
        }

        @Override
        public CustomerEphemeralKey[] newArray(int size) {
            return new CustomerEphemeralKey[size];
        }
    };

    private CustomerEphemeralKey(@NonNull Parcel in) {
        super(in);
    }

    private CustomerEphemeralKey(long created, @NonNull String customerId, long expires,
                                 @NonNull String id, boolean liveMode, @NonNull String object,
                                 @NonNull String secret, @NonNull String type) {
        super(created, customerId, expires, id, liveMode, object, secret, type);
    }

    @NonNull
    String getCustomerId() {
        return mObjectId;
    }

    @NonNull
    static CustomerEphemeralKey fromJson(@NonNull JSONObject jsonObject) throws JSONException {
        return EphemeralKey.fromJson(jsonObject, new Factory());
    }

    static final class Factory extends EphemeralKey.Factory<CustomerEphemeralKey> {
        @NonNull
        @Override
        CustomerEphemeralKey create(long created, @NonNull String objectId, long expires,
                                       @NonNull String id, boolean liveMode, @NonNull String object,
                                       @NonNull String secret, @NonNull String type) {
            return new CustomerEphemeralKey(created, objectId, expires, id, liveMode, object,
                    secret, type);
        }
    }
}

File: stripe/src/main/java/com/stripe/android/CustomerEphemeralKey.kt
package com.stripe.android

import android.os.Parcel
import android.os.Parcelable
import org.json.JSONException
import org.json.JSONObject

internal class CustomerEphemeralKey : EphemeralKey {
    val customerId: String
        get() = objectId

    private constructor(parcel: Parcel) : super(parcel)

    private constructor(
        created: Long,
        customerId: String,
        expires: Long,
        id: String,
        liveMode: Boolean,
        objectType: String,
        secret: String,
        type: String
    ) : super(created, customerId, expires, id, liveMode, objectType, secret, type)

    internal class Factory : EphemeralKey.Factory<CustomerEphemeralKey>() {
        override fun create(
            created: Long,
            objectId: String,
            expires: Long,
            id: String,
            liveMode: Boolean,
            objectType: String,
            secret: String,
            type: String
        ): CustomerEphemeralKey {
            return CustomerEphemeralKey(
                created, objectId, expires, id, liveMode, objectType, secret, type
            )
        }
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<CustomerEphemeralKey> =
            object : Parcelable.Creator<CustomerEphemeralKey> {

                override fun createFromParcel(`in`: Parcel): CustomerEphemeralKey {
                    return CustomerEphemeralKey(`in`)
                }

                override fun newArray(size: Int): Array<CustomerEphemeralKey?> {
                    return arrayOfNulls(size)
                }
            }

        @Throws(JSONException::class)
        @JvmStatic
        fun fromJson(jsonObject: JSONObject): CustomerEphemeralKey {
            return fromJson(jsonObject, Factory())
        }
    }
}
